### PR TITLE
Task alias validation during training mode

### LINKED ISSUE #740: --task transformer doesn't do anything and gives somewhat suboptimal defaults
### Bug description
It used to be the case that we had two separate `--task` possibilities: `transformer` and `transformer-big` and this is still reflected in the options ( https://github.com/marian-nmt/marian-dev/blob/master/src/common/config_parser.cpp#L536 ) or on the website.

This is supposed to be matched by an alias entry ( https://github.com/marian-nmt/marian-dev/blob/master/src/common/aliases.cpp#L74 ) but it got renamed to `transformer-base`.  As a result, new users end up putting `--task transformer` and are getting  some largely unexpected defaults.

We should update the help description and also warn/crash when unknown task is provided, so that the user knows when they are using outdated/wrong parameters.

Cheers,

Nick

### PR DESCRIPTION
**Pull request by XapaJIaMnu**
### Description
This PR fixes #740 by adding an explicit validation of the task alias. Beforehand it used to be the case that if a task alias was provided that is not in `aliases.cpp` you would get random default options and things would break in unexpected ways with no feedback.

List of changes:
- Added validation of the task option if the mode is training.
- Updated the help at `config_parser.cpp`
- Removed the behaviour where aliases are removed from the config after being parsed so that we can check them later. Not sure if this breaks anything else. @graemenail  is there a better way to achieve this?

Added dependencies: none

### How to test
If you provide something like `task: transformer` you will get something like:
```
[2021-10-25 15:25:47] Error: Unrecognised task option: "transformer". Permited "transformer-base-prenorm" "transformer-big-prenorm" "transformer-big" "transformer-base" "" 
[2021-10-25 15:25:47] Error: Aborted from int mainTrainer(int, char**) in /home/USER/marian-dev/src/command/marian_train.cpp:26
```

### Checklist

- [x] I have tested the code manually
- [x] I have run regression tests
- [x] I have read and followed CONTRIBUTING.md
- [x] I have updated CHANGELOG.md

### CODE REVIEW

**Comment by snukky** [comment]
@XapaJIaMnu I made some changes to your code. Can you take a look if my proposed solution does exactly what you initially wanted to change with this PR? Thanks!

I tested it manually, added a simple regression test (https://github.com/marian-nmt/marian-regression-tests/pull/82), and run regression tests again.

File: src/common/cli_wrapper.cpp [hunk 1]
```diff
@@ -132,6 +132,12 @@ void CLIWrapper::parseAliases() {
   if(aliases_.empty())
     return;
 
+  // Find the set of values allowed for each alias option.
+  // Later we will check and abort if an alias option has an unknown value.
+  std::unordered_map<std::string, std::unordered_set<std::string>> allowedAliasValues;
+  for(const auto &alias : aliases_)
```
Comment (XapaJIaMnu) [inline src/common/cli_wrapper.cpp:138]: I prefer `auto&&` here to make sure it doesn't copy (but that's mostly an issue with 3rd party container that have poorly defined iterators... Doesn't make a difference here.

File: src/common/cli_wrapper.cpp [hunk 2]
```diff
@@ -145,6 +151,12 @@ void CLIWrapper::parseAliases() {
       bool expand = false;
       if(config_[alias.key].IsSequence()) {
         auto aliasOpts = config_[alias.key].as<std::vector<std::string>>();
+        // Abort if an alias option has an unknown value, i.e. value that has not been defined
+        // in common/aliases.cpp
+        for(const auto &aliasOpt : aliasOpts)
```
Comment (XapaJIaMnu) [inline src/common/cli_wrapper.cpp:156]: Same as above.

File: src/common/cli_wrapper.cpp [hunk 3]
```diff
@@ -145,6 +151,12 @@ void CLIWrapper::parseAliases() {
...
+          if(allowedAliasValues[alias.key].count(aliasOpt) == 0)
+            ABORT("Unknown value '" + aliasOpt + "' for alias option --" + alias.key + ". "
+                  "Run with --help to see allowed values.");
```
Comment (XapaJIaMnu) [inline src/common/cli_wrapper.cpp:159]: Maybe just print the allowed values here, since we have access to them?

File: src/common/cli_wrapper.cpp [hunk 4]
```diff
@@ -145,6 +151,12 @@ void CLIWrapper::parseAliases() {
       bool expand = false;
       if(config_[alias.key].IsSequence()) {
         auto aliasOpts = config_[alias.key].as<std::vector<std::string>>();
+        // Abort if an alias option has an unknown value, i.e. value that has not been defined
+        // in common/aliases.cpp
+        for(const auto &aliasOpt : aliasOpts)
+          if(allowedAliasValues[alias.key].count(aliasOpt) == 0)
```
Comment (XapaJIaMnu) [inline src/common/cli_wrapper.cpp:157]: I personally prefer to use 
`if(allowedAliasValues.find(alias.key) == allowedAliasValues::end())`

**Review by XapaJIaMnu — COMMENTED** [review]
So, I can't approve the pull request, because I can't approve my own pull request xD.

I think what you have done works, some stylistic suggestions based on what code i like to write, but i don't think it matters. The only suggestion is that we should print all available aliases when there is a mistake, as opposed to asking the user to rerun with `--help`.

File: src/common/cli_wrapper.cpp (hunk 1 above)
Comment (snukky) [reply src/common/cli_wrapper.cpp]: Changed, thanks!

File: src/common/cli_wrapper.cpp (hunk 3 above)
Comment (snukky) [reply src/common/cli_wrapper.cpp]: Added.

File: src/common/cli_wrapper.cpp (hunk 4 above)
Comment (snukky) [reply src/common/cli_wrapper.cpp]: It would be `if(allowedAliasValues[alias.key].find(aliasOpt) == allowedAliasValues[alias.key].end())` and I kept the previous version because it was shorter.

**Review by snukky — APPROVED** [review]
